components/brave_news/browser/signal_calculator.cc

```diff
@@ -0,0 +1,192 @@
+      history_service_(history_service) {}
+
+SignalCalculator::~SignalCalculator() = default;
+
+void SignalCalculator::GetSignals(const FeedItems& feed,
+                                  SignalsCallback callback) {
+  auto articles = GetArticles(feed);
+  history::QueryOptions options;
+  options.SetRecentDayRange(21);
+  options.max_count = 2000;
+  history_service_->QueryHistory(
+      u"", options,
+      base::BindOnce(&SignalCalculator::OnGotHistory,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(articles),
+                     std::move(callback)),
+      &task_tracker_);
+}
+
+void SignalCalculator::OnGotHistory(
+    std::vector<mojom::FeedItemMetadataPtr> articles,
+    SignalsCallback callback,
+    history::QueryResults results) {
+  const auto& locale = publishers_controller_->GetLastLocale();
+
+  auto& publishers = publishers_controller_->GetLastPublishers();
+  auto channels = channels_controller_->GetChannelsFromPublishers(
+      publishers, &prefs_.get());
+  base::flat_map<std::string, std::vector<std::string>> origin_visits;
+  for (const auto& item : results) {
+    auto host = item.url().host();
+    origin_visits[host].push_back(item.url().spec());
+  }
+
+  // Start at one - it'll make the calculations very slightly off but it also
+  // means we'll never divide by zero, and it will be consistent.
+  uint32_t total_publisher_visits = 1;
+  uint32_t total_channel_visits = 1;
+
+  base::flat_map<std::string, std::vector<std::string>> publisher_visits;
+  base::flat_map<std::string, std::vector<std::string>> channel_visits;
+
+  for (auto& [publisher_id, publisher] : publishers) {
+    auto host = publisher->site_url.host();
+    auto history_it = origin_visits.find(host);
+    if (history_it == origin_visits.end()) {
+      publisher_visits[publisher_id] = {};
+      continue;
+    }
+
+    base::ranges::copy(history_it->second,
+                       std::back_inserter(publisher_visits[publisher_id]));
+    total_publisher_visits += history_it->second.size();
+
+    for (const auto& locale_info : publisher->locales) {
+      if (locale_info->locale != locale) {
+        continue;
+      }
+
+      for (const auto& channel : locale_info->channels) {
+        total_channel_visits += history_it->second.size();
+        base::ranges::copy(history_it->second,
+                           std::back_inserter(channel_visits[channel]));
+      }
+      break;
+    }
+  }
+
+  Signals signals;
+
+  // Add article signals
+  for (const auto& article : articles) {
+    const auto& publisher = publishers.at(article->publisher_id);
+
+    bool channel_subscribed = false;
+    for (const auto& locale_info : publisher->locales) {
+      for (const auto& channel : locale_info->channels) {
+        if (channels_controller_->GetChannelSubscribed(locale_info->locale,
+                                                       channel)) {
+          channel_subscribed = true;
+          break;
+        }
+      }
+    }
```

Comment:
How about moving this part into IsPublisherSubscribe() and make it take Publisher as parameter? I hope that could make building signals more concise.